// @wdio/reporter/build/stats/test.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _runnable = _interopRequireDefault(require("./runnable"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * TestStats class
 * captures data on a test.
 */
class TestStats extends _runnable.default {
  constructor(runner) {
    super('test');
    this.uid = _runnable.default.getIdentifier(runner);
    this.cid = runner.cid;
    this.title = runner.title;
    this.fullTitle = runner.fullTitle;
    this.output = [];
    /**
     * initial test state is pending
     * the state can change to the following: passed, skipped, failed
     */

    this.state = 'pending';
  }

  pass() {
    this.complete();
    this.state = 'passed';
  }

  skip() {
    this.state = 'skipped';
  }

  fail(error) {
    this.complete();
    this.state = 'failed';
    this.error = error;
  }

}

exports.default = TestStats;
